master/internal/provisioner/scale_decider.go

```diff
@@ -30,26 +29,29 @@ type scaleDecider struct {
 	maxIdlePeriod       time.Duration
 	maxStartingPeriod   time.Duration
 	maxDisconnectPeriod time.Duration
+	maxInstanceNum      int
 
 	lastProvision          time.Time
 	lastSchedulerUpdated   time.Time
 	lastProviderUpdated    time.Time
 	instanceSnapshot       map[string]*Instance
-	idleAgentSnapshot      map[string]*scheduler.AgentSummary
-	connectedAgentSnapshot map[string]*scheduler.AgentSummary
-	taskSnapshot           []*scheduler.TaskSummary
+	idleAgentSnapshot      map[string]sproto.AgentSummary
+	connectedAgentSnapshot map[string]sproto.AgentSummary
+
+	desiredNewInstanceNum int
```

Comment:
I would maybe change this to `desiredNumInstances` or something. `desiredNewInstanceNum` is a little vague on whether this is the total number of instances we need or the number of new instances that we need to launch.